cuda_core/cuda/core/experimental/_device.py

```diff
@@ -957,26 +957,34 @@ def __new__(cls, device_id=None):
 
         # important: creating a Device instance does not initialize the GPU!
         if device_id is None:
-            device_id = handle_return(runtime.cudaGetDevice())
-            assert_type(device_id, int)
+            err, dev = driver.cuCtxGetDevice()
+            if err == 0:
+                device_id = int(dev)
+            else:
+                ctx = handle_return(driver.cuCtxGetCurrent())
+                assert int(ctx) == 0
+                device_id = 0  # cudart behavior
         else:
-            total = handle_return(runtime.cudaGetDeviceCount())
-            assert_type(device_id, int)
-            if not (0 <= device_id < total):
+            total = handle_return(driver.cuDeviceGetCount())
+            if not isinstance(device_id, int) or not (0 <= device_id < total):
                 raise ValueError(f"device_id must be within [0, {total}), got {device_id}")
 
         # ensure Device is singleton
-        if not hasattr(_tls, "devices"):
-            total = handle_return(runtime.cudaGetDeviceCount())
-            _tls.devices = []
+        try:
+            devices = _tls.devices
+        except AttributeError:
+            total = handle_return(driver.cuDeviceGetCount())
+            devices = _tls.devices = []
             for dev_id in range(total):
```

Comment:
If someone tries to create a Device with a specific ID, why do we need to initialize all of the devices at that point? Instead of calling `driver.cuDeviceGetCount()` to get the number of devices to initialize all of `_tls.devices`, could we use `driver.cuDeviceGet` and lazily populate `_tls.devices` as devices are created?

Reply:
We had discussions about this code back in March. @leofang [wrote here](https://github.com/NVIDIA/cuda-python/pull/459#issuecomment-2726176716):

> Each thread always has its own copy of `_tls.devices`.

I'm still unclear though TBH: Is that why we cannot lazily populate?

Reply:
Each thread has a private storage of an N-list (N is the number of devices seen in the process), and currently the list is initialized at once instead of lazily. Since this is a one-time only initialization we're discussing here, I prefer to pay this cost upfront, to reduce the possibility of perf jittering in a multithreading code. Given the ongoing perf investigation (#677) we might end up dropping all of these later. So let's keep it simple for now.